### PR TITLE
WIP: Enhancement proposal of SuccessiveHalvingPruner

### PR DESCRIPTION
**Pull request by c-bata**
## Preface

SuccessiveHalvingPruner take different behaviors depending on the order of trial's performances.  In #388, it pruned a trial which is 2nd-best score (and 23/25 trials are pruned at rung 0). On the other hand, as [this benchmark](https://gist.github.com/sile/d4befbe5a3b1ec9711cb5dcd5dbc4538) shows the average of required budgets are relatively expensive. These results shows current implementation is completely different algorithm from the original paper.

The reason why SuccessiveHalvingPruner behaves like this is that it's difficult to provide user-friendly api to saving/restoring trained models of each trials at the increases of the rung. So optuna couldn't implement exactly the same behavior with the original paper.

## Proposal

My proposal is simple, when the rung increases, it suspends the running trial without saving trained models, then start training the model over from the beginning.

This proposal looks inefficient because of the waste of training. But it's better not only this proposal is robust against the performance order of each trials but also it's better in the viewpoint of calculation costs.

### Does my proposal makes the total required budgets bigger?

No. The step count of my proposal is fewer than current SuccessiveHalvingPruner.

| reduction_factor | SHA steps | ASHA steps (SD) | Optuna steps (SD) | Proposal method(SD) |
|------------------:|------------:|--------------:|----------------:|------------:|
| 2                        | 376           |  564 (74)       | 1363 (603)        | 1099 (171) |
| 3                        |  340          | 445 (65)        | 818 (275)        |  654 (129) |
| 4                        |  295          | 389 (66)        | 763 (243)        | 496 (104)  |

The benchmark is [here](https://gist.github.com/c-bata/b3ec610b896533729846059bb42459f1), which is based on [@sile's benchmark](https://gist.github.com/sile/d4befbe5a3b1ec9711cb5dcd5dbc4538).

The required budgets on each rung is exponentially increased with increases of the rung. Optuna always consumes 100 steps to evaluate first trial in the benchmark. This budget equals how much we need to execute rung 0 on all trials. We need to estimate the cost of proceeding the rung to the end more expensive. This is the reason why step count of my proposal is fewer than the ones of the current implementation.

The another advantage of this method is low standard deviation. This means we can easily expect the execution time of optimization.

###  Can we provide APIs to save/restore trained models?

Probably yes. There are several approaches, But I have no definitive solutions. It seems we need to discuss more. Now I think following interface might be good in the viewpoint of usability.

The most important thing is we can use SuccessiveHalvingPruner even if we don't use these apis. I think it should be optional.

### Is it Asynchronous Successive Halving Algorithm?

Yes. The original paper of 'Massively Parallel Hyperparameter Optimization' doesn't mention whether we should suspend/resume trials. The GIF animation in the blog post of Liam Li (1st author of the ASHA's paper) behaves the same with my proposal.

> asha
> 
> https://blog.ml.cmu.edu/2018/12/12/massively-parallel-hyperparameter-optimization/

## Discussion

I think there are no disadvantages very much in this proposal, but it seems we need to discuss the design of each components to keep clean (ex:  pruner's interface). This PR is still WIP. I want your feedbacks on this proposal before the implementation is ready for review.

Another things I want to consider.

* How we provide the apis to suspend training. Should we provide `should_suspend` method and `TrialSuspended` ?
* It might be better to define Interface for storage backend for serialized models. We may not want to use RDB because the parameters of modern machine learning models (like deep neural-network) tends to be bigger.
* It might be better to benchmark this method by more realistic problems.

## Implementation status

This PR depends #464, ~~#424~~ , #493 and #520.

### CODE REVIEW

**Comment by sile** [comment]
Thank you for your proposal.
I think your approach is very good because it faithfully reflects the description of the original paper 
(honestly, I was considering a similar reimplementation of `SuccessiveHalvingPruner`).

However, as you mentioned in the "Discussion" section, it requires some design decisions and benchmarking for merging into the master branch.
We would like to discuss this more deeply within Optuna dev team within a few weeks, so please give us some time.

**Comment by c-bata** [comment]
Thank you! I'm looking forward to hearing your decision :+1: This improvements is much needed for me.

**Comment by sile** [comment]
Sorry for the late reply.

First of all, I write a summary of Optuna team's discussion result.

We suggest, 
- (a) Removing snapshot feature (saving/loading models) from the scope of this PR
- (b) Keeping the existing algorithm and adding a flag to switch between it and the new one
- (c) Not introducing `should_suspend` method and reusing `should_prune` for this purpose instead
- (d) Using `WAITING` state (that is going to be added by #424) instead of `SUSPEND` and `PROMOTABLE`

Please refer to the following sections for the details.

### (a) Removing snapshot feature (saving/loading models) from the scope of this PR

We think it's great to support saving/loading models. However, there are many things to consider in order to realize it. For keeping this PR simple, we want to drop the snapshot feature from this PR and work on it in the future.

### (b) Keeping the existing algorithm and adding a flag to switch between it and the new one

I ran some benchmarks to estimate the performance of the new ASHA implementation.
The figure below is one of the benchmark results (see [the gist](https://gist.github.com/sile/78500d7103c84f35d2c1cb2850abe564) for full results):

MFB1

In this figure, "Budget" means the total number of steps consumed by an optimizer (pruner) and "Object Value" means the best objective value found by an optimizer within the budget.
There are three optimizers:
- `optuna-asha`: The current implementation of `SuccessiveHalvingPruner`
- `paper-asha-with-snapshot`: An implementation of ASHA that faithfully reflects the paper
- `paper-asha-without-snapshot`: The same as `paper-asha-with-snapshot` but this doesn't assume snapshot capability

The following points can be seen from this result (the other benchmarks also show similar results):
- `optuna-asha` tends to lose to others when the budget is small:
  - But, if there is enough budget, it achieves almost the same performance with `paper-asha-with-snapshot`
- Due to the overhead of redundant re-execution, `paper-asha-without-snapshot` tends to be the worst performance

There are pros and cons in the original algorithm and the Optuna's algorithm, and it doesn't seem that one is completely superior to the other:

|                             | Pros | Cons |
|-----------------------------|------|------|
| optuna-asha                 | shows good performance if there is enough budget, doesn't require snapshot | shows worse performance if the budget is small |
| paper-asha-with-snapshot    | shows the best performance | requires snapshot | 
| paper-asha-without-snapshot | doesn't require snapshot | shows worse performance than `paper-asha-with-snapshot`  |

With this result in mind, we would like to keep the existing algorithm (`optuna-asha`) and add a flag to switch between it and the new one.

### (c) Not introducing `should_suspend` method and reusing `should_prune` for this purpose instead

We would like to re-use the existing `should_prune` method for the following reasons:
- Adding `should_suspend` method may cause confusion for users and implementors of pruners
  - E.g., "There are `should_suspend` and `should_prune`, but which one should I use?"
- I think that the equivalent of suspend/resume can be realized in the existing framework without introducing new API to `BasePruner`:
  - Please see the next section for the details
- Considering to support snapshot, we may need to introduce the concept of "suspension"
  - However, as mentioned above, I think that it should be considered separately from this PR

### (d) Using `WAITING` state (that is going to be added by #424) instead of `SUSPEND` and `PROMOTABLE`

We plan to introduce "trial queue" feature by #424.
This is a feature that allows users to enqueue trials that have `WAITING` state and user-specified parameters. In a round of an optimization process, if there are `WAITING` trials, a study runs one of those trials instead of creating a new trial.

I think that the paper's ASHA algorithm can be implemented by using this feature as follows:
1. Transit the state of a trial that has completed the current rung to `PRUNED`
2. Determine the trial to promote to the next rung according to the ASHA algorithm
3. Copy the trial parameters and system attributes and enqueue it to the trial queue
   - After that, the trial queue mechanism can be used as it is

In this way, I think that ASHA can be implemented using the more general trial queue mechanism without adding the states `SUSPEND` and `PROMOTABLE` (as a disadvantage of the way, the ID of a trial changes every time it's promoted).

Although the PR has not been created yet, we plan to extend the mechanism further to support changing the existing trial state to `WAITING` so that it can be enqueued again after (for example) pruned.
(I think this is almost the same as the feature you are trying to add by this PR except for the name.)

---

Please let us know if you have any thoughts about this suggestion.

**Comment by c-bata** [comment]
Thank you for sharing your considerations. It sounds your opinions are completely reasonable.
And your benchmark experiment results are very interesting for me.

One thing I'm concerned about is that I don't know it's better to switch between `optuna-asha` and `paper-asha-without-snapshot` by a flag. I think we should consider to implement `paper-asha-without-snapshot` as an another Pruner class.

* If SuccessiveHalvingPruner switches `optuna-asha` or `paper-asha-without-snapshot` by a flag, we need to decide which should be default behavior.
* If we will implement it as another class, we need to decide whether the SuccessiveHalvingPruner class should be `optuna-asha` or `paper-asha-without-snapshot`. I think optuna-asha may be named another name as new algorithm, not ASHA.

**Comment by sile** [comment]
@c-bata Thank you for your reply.

> I don't know it's better to switch between `optuna-asha` and `paper-asha-without-snapshot` by a flag.
> I think we should consider to implement `paper-asha-without-snapshot` as an another Pruner class.

Though `oputna-asha` is not an exact implementation of [ASHA], both of `optuna-asha` and `paper-asha-without-snapshot` belong to the same family of successive halving algorithms. Therefore, we think it's reasonable to provide both via `SuccessiveHalvingPruner` class.

> we need to decide which should be default behavior.

Considering [my benchmark result](https://gist.github.com/sile/78500d7103c84f35d2c1cb2850abe564), we think it would be better to use the current implementation (`optuna-asha`) as the default.

[ASHA]: https://arxiv.org/abs/1810.05934

**Comment by sile** [comment]
IMO, only two simple flags may be sufficient to switch the behaviors of both algorithms.

There aren't so much differences between `optuna-asha` and `paper-asha-without-snapshot`.
The essential differences are,

- (a) whether the algorithm allows to restart pruned trials later
   - `optuna-asha` prohibits this, and `paper-asha-without-snapshot` allows this
- (b) whether the algorithm allows to promote trials even if the number of trials in a rung is insufficient
   - `optuna-asha` allows this, and `paper-asha-without-snapshot` prohibits this

**Comment by c-bata** [comment]
I see! I implement this feature in the SuccessiveHalvingPruner, and switches it by a flag.
